Fix readjustment of number of slices bug in RSA

Rosejoycrocker opened the pull request:
Fixes readjustment of number of slices bug in RSA (partially addresses #615) and changes output of RSA to `YAXArrays.Dataset` type.

ConnectedSystems commented:
Does the switch to YAXArray break the ability to plot results? 
I don't see any compatibility layer in place.

Rosejoycrocker commented:
> Does the switch to YAXArray break the ability to plot results? I don't see any compatibility layer in place.

There are also changes to `viz.rsa()` in this PR so plotting works

ConnectedSystems commented on ext/AvizExt/viz/viz.jl:
```diff
@@ -1,7 +1,7 @@
 
 using ADRIA: ResultSet, metrics.metric_label, analysis.col_normalize, model_spec
-using NamedDims, AxisKeys
+using NamedDims, AxisKeys, YAXArrays, DimensionalData
```
Is the import of DimensionalData simply to have the `At` method?

If so, I'd prefer it be imported via YAXArrays:

```julia
import YAXArrays.DD: At
```

Rosejoycrocker replied:
I think I just took this from the YAXArray docs and forgot to remove, I've removed it now

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -5,7 +5,9 @@ using Logging
     NamedDims,
-    StaticArrays
+    StaticArrays,
+    YAXArrays,
+    DimensionalData
```
As with previous comment - if this is for `At`, import it directly.

ConnectedSystems commented:
> > Does the switch to YAXArray break the ability to plot results? I don't see any compatibility layer in place.
> 
> There are also changes to `viz.rsa()` in this PR so plotting works

Thanks Rose, clearly I'm going blind.

ConnectedSystems commented on ext/AvizExt/viz/sensitivity.jl:
```diff
@@ -175,7 +175,7 @@ Makie figure
     g::Union{GridLayout,GridPosition},
     rs::ResultSet,
-    si::NamedDimsArray,
+    si::YAXArrays.Dataset,
```
Why not import `Dataset` directly so prefacing it with `YAXArrays` is not necessary?

Or is there a different `Dataset` type that is causing a conflict?

Rosejoycrocker replied:
I've replaced with `Dataset`

ConnectedSystems commented on src/analysis/sensitivity.jl:
```diff
@@ -451,78 +455,107 @@ ADRIA.sensitivity.rsa(X, y; S=10)
     N, D = size(X)
 
-    X_di = zeros(N)
+    X_i = zeros(N)
```
There was a reason for X_di ($X_{d, i}$) but that may now be lost to the sands of time...

Rosejoycrocker replied:
Ah makes sense. Should I revert or leave as is?

ConnectedSystems replied:
Leave as is

Rosejoycrocker commented:
> Some minor comments for consideration

I think I've addressed your comments, If you're happy with this I'll do another PR with similar changes for `outcome_map`

ConnectedSystems commented:
Busy with something else at the moment, would you mind attaching a figure here to demonstrate its working please?

Rosejoycrocker commented:
> Busy with something else at the moment, would you mind attaching a figure here to demonstrate its working please?

Here's an example 
![sensitivity_rsa_ex](https://github.com/open-AIMS/ADRIA.jl/assets/56939532/0e2b6d00-bd32-4b6d-a779-03de11a0c1cb)

ConnectedSystems commented:
The peak is telling you that there is high variability in that factor space.

In other words, the outcome assessed is highly sensitive to that region of factor space, which corresponds with no intervention.

Rosejoycrocker commented:
> The peak is telling you that there is high variability in that factor space.
> 
> In other words, the outcome assessed is highly sensitive to that region of factor space, which corresponds with no intervention.

Makes sense

ConnectedSystems reviewed (approved):
LGTM
